### PR TITLE
feat: introduce Rich Text Editor component

### PR DESCRIPTION
**Pull request by manu-kr**
## Related Issue
Closes https://github.com/SAP/fundamental-styles/issues/2198

## Description
Introducing new rich text editor is component.

### After:
image

#### Please check whether the PR fulfills the following requirements

1. The output matches the design specs
- [x] All values are in `rem`
- [x] Text elements follow the truncation rules
- [x] hover state of the element follow design spec
- [x] focus state of the element follow design spec
- [x] active state of the element follow design spec
- [x] selected state of the element follow design spec
- [x] selected hover state of the element follow design spec
- [x] pressed state of the element follow design spec
- [x] Responsiveness rules - the component has modifier classes for all breakpoints
- [x] Includes Compact/Cosy/Tablet design
- [x] RTL support
2. The code follows fundamental-styles code standards and style
- [NA] only one top level `fd-*` class is used in the file
- [NA] BEM naming convention is used
- [NA] Mixins are used for repeatable code (`fd-rtl`, `fd-ellipsis`, `fd-flex`, `fd-selected`, `fd-focus`, ect.)
- [x] A11y support - keyboard support, screenreader support, proper ARIA attributes, etc.
- [NA] `fd-reset()` mixin is applied to all elements
- [NA] Variables are used, if some value is used more than twice.
- [NA] Checked if current components can be reused, instead of having new code.
3. Testing
- [x] Verified all styles in IE11
- [x] Updated tests
- [x] last commit message should have `[ci visual]` so it can trigger chromatic visual regression (e.g. `test: run chromatic visual regression [ci visual]`)
4. Documentation
- [x] Storybook documentation has been created/updated
- [ NA] Breaking Changes [wiki](https://github.com/SAP/fundamental-styles/wiki/Breaking-Changes) has been updated in case of breaking changes.

### CODE REVIEW

**Comment by InnaAtanasova** [comment]
1. Problem with the active state - the letter A should be white
Screen Shot 2021-03-24 at 1 46 22 PM

2. You should have 1 example in the docs where A (Title Toolbar) and B (Toolbar with overflow) have white background as in the specs:
Screen Shot 2021-03-24 at 1 45 23 PM

File: stories/rich-text-editor/rich-text-editor.stories.js [hunk 1]
```diff
@@ -0,0 +1,852 @@
...
+     <i class="sap-icon--bullet-text"></i>
+    </button>
+
+    <button class="fd-button fd-button--transparent" aria-label="numbered text">
+    <i class="sap-icon--numbered-text"></i>
+   </button>
+
+</div>
+<textarea aria-label="rtecozytextarea" class="fd-textarea" cols="30" rows="15" id="textarea-2">Pellentesque metus lacus commodo eget justo ut rutrum varius nunc.</textarea>
+</div>
```
[the remark this answers is not in the document]
Comment (kavya-b) [reply stories/rich-text-editor/rich-text-editor.stories.js:281]: also, same id for textarea given in other places

**Review by kavya-b — CHANGES_REQUESTED** [review]
Few comments, please take a look:
- Quartz Dark, `A` text button is not visible
image
- is the color palette not going to be implemented? 
image

**Comment by manu-kr** [comment]
> Few comments, please take a look:
> 
> * Quartz Dark, `A` text button is not visible
>   image
> * is the color palette not going to be implemented?
>   image

Yes it's a another fiori3 component has to implement.

**Comment by manu-kr** [comment]
> 1. Problem with the active state - the letter A should be white
> 
> Screen Shot 2021-03-24 at 1 46 22 PM
> 
> 1. You should have 1 example in the docs where A (Title Toolbar) and B (Toolbar with overflow) have white background as in the specs:
> 
> Screen Shot 2021-03-24 at 1 45 23 PM

Addressed

File: stories/rich-text-editor/rich-text-editor.stories.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (manu-kr) [reply stories/rich-text-editor/rich-text-editor.stories.js:281]: Modified.

File: stories/rich-text-editor/rich-text-editor.stories.js [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+                            <span class="fd-button fd-button--transparent fd-select__button">
+                                <i class="sap-icon--slim-arrow-down"></i> 
+                            </span>
+                        </button>
+                    </div>
+                </div>
+        <div aria-hidden="true" class="fd-popover__body fd-popover__body--no-arrow fd-popover__body--dropdown" id="h0C6A325">
+            <ul aria-activedescendant="cozyRTEParagraph-currentlyFocusedItem" aria-label="cozyselectlabel" class="fd-list fd-list--dropdown" role="listbox">
+                <li id="cozyRTEParagraph-currentlyFocusedItem" class="fd-list__item" role="option" tabindex="0">
+                <span class="fd-list__title">Select</span>
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply stories/rich-text-editor/rich-text-editor.stories.js:84]: Modified

File: stories/rich-text-editor/rich-text-editor.stories.js [hunk 3]
```diff
@@ -0,0 +1,852 @@
...
+                    
+                    </ul>
+                </div>
+        </div>
+
+
+        <div class="fd-popover">
+                <div class="fd-popover__control">
+                    <div class="fd-select">
+                        <button aria-expanded="false" aria-haspopup="listbox" aria-label="cozyselectlabel cozyselectvalue" class="fd-select__control" id="cozyRTEParagraph" onclick="
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply stories/rich-text-editor/rich-text-editor.stories.js:70]: modified.

File: stories/rich-text-editor/rich-text-editor.stories.js [hunk 4]
```diff
@@ -0,0 +1,852 @@
...
+                                        toggleElAttrs('h0C6A325', ['aria-hidden']);
+                                        toggleElAttrs('cozyRTEParagraph', ['aria-expanded']);
+                                    " value="List Item 1" tabindex="0">
+                            <span is="cozyselectvalue" class="fd-select__text-content">Paragraph</span>
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply stories/rich-text-editor/rich-text-editor.stories.js:74]: modified.

File: stories/rich-text-editor/rich-text-editor.stories.js [hunk 5]
```diff
@@ -0,0 +1,852 @@
...
+                        <button aria-expanded="false" aria-haspopup="true" aria-label="cozy rte allignments"
+                        class="fd-button fd-button--menu fd-button--transparent" 
+                        aria-controls="h0C6A324"
+                        tabindex="0">
+                                <i class="sap-icon--text-align-left"></i>
+                                <i class="sap-icon--slim-arrow-down"></i> 
+                        </button>
+                </div>
+                <div aria-hidden="true" class="fd-popover__body fd-popover__body--no-arrow fd-popover__body--dropdown" id="h0C6A324">
+                    <ul aria-activedescendant="cozySelectRTE-currentlyFocusedItem" aria-labelledby="cozyRTELabel" class="fd-list fd-list--dropdown" role="listbox">
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply stories/rich-text-editor/rich-text-editor.stories.js:45]: modified.

**Comment by manu-kr** [comment]
> 1. Problem with the active state - the letter A should be white
> 
> Screen Shot 2021-03-24 at 1 46 22 PM
> 
> 1. You should have 1 example in the docs where A (Title Toolbar) and B (Toolbar with overflow) have white background as in the specs:
> 
> Screen Shot 2021-03-24 at 1 45 23 PM
There is no option to have white background for toolbar.

File: stories/rich-text-editor/__snapshots__/rich-text-editor.stories.storyshot [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Storyshots Patterns/Rich Text Editor Compact Mode 1`] = `
```
Comment (kavya-b) [inline stories/rich-text-editor/__snapshots__/rich-text-editor.stories.storyshot:3]: storyshots have become empty, please fix this. Usually this is due to formatting differences. Please check some of the other components where storyshots are fine and try similar changes.

File: stories/rich-text-editor/rich-text-editor.stories.js [hunk 7]
```diff
@@ -0,0 +1,1116 @@
...
+			</div>
+		    </div>
+		    <div class="fd-popover">
+			<div class="fd-popover__control">
+			    <div class="fd-select fd-select--compact">
+				<button aria-expanded="false" aria-haspopup="listbox" aria-label="select-label" class="fd-select__control" id="overflowRTEFontSize" onclick="
+				    toggleElAttrs('h0C6A3217', ['aria-hidden']);
+				    toggleElAttrs('overflowRTEFontSize', ['aria-expanded']);
+				    " value="List Item 1" tabindex="0">
+				<span is="overflowSelectValue" class="fd-select__text-content">10 pt</span>
```
Comment (kavya-b) [inline stories/rich-text-editor/rich-text-editor.stories.js:995]: `id` typo? Please check in other places too

**Review by kavya-b — CHANGES_REQUESTED** [review]
Few comments. Also please check all relevant boxes in the PR description.
- In only the first example, `A` font appears fine in themes. The other examples don't show it correctly: 

image

File: stories/rich-text-editor/rich-text-editor.stories.js [hunk 8]
```diff
@@ -0,0 +1,852 @@
+export default {
+    title: 'Patterns/Rich Text Editor',
+    parameters: {
+        description: `
+The rich-text-editor component is an opinionated composition of the \`toolbar\`, \`toggle-button\`, \`menu-button\`, \`split-button\`, \`select\`,\`icon\`,\`popover\`,\`menu\`,\`list\` and \`text-area\` components.
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply stories/rich-text-editor/rich-text-editor.stories.js]: modified.

File: stories/rich-text-editor/rich-text-editor.stories.js (hunk 7 above)
Comment (manu-kr) [reply stories/rich-text-editor/rich-text-editor.stories.js]: modified.

File: stories/rich-text-editor/__snapshots__/rich-text-editor.stories.storyshot (hunk 6 above)
Comment (manu-kr) [reply stories/rich-text-editor/__snapshots__/rich-text-editor.stories.storyshot:3]: updated storyshot.

**Comment by kavya-b** [comment]
> Few comments. Also please check all relevant boxes in the PR description.
> 
> * In only the first example, `A` font appears fine in themes. The other examples don't show it correctly:
> 
> image

@manu-kr these comments have not been addressed.

**Comment by manu-kr** [comment]
> ## Related Issue
> Closes #2198
> 
> ## Description
> Introducing new rich text editor is component.
> 
> ## Screenshots
> > **NOTE:** If you've made any style changes, please provide appropriate screenshots (before and after) to help reviewers.
> > To see examples of which screenshots to include, go to [Screenshot Examples](https://github.com/SAP/fundamental-styles/wiki/Pull-Request-Screenshot-Examples).
> 
> ### Before:
> ### After:
> image
> 
> #### Please check whether the PR fulfills the following requirements
> 1. The output matches the design specs
> 
> * [X] All values are in `rem`
> * [X] Text elements follow the truncation rules
> * [X] hover state of the element follow design spec
> * [X] focus state of the element follow design spec
> * [X] active state of the element follow design spec
> * [X] selected state of the element follow design spec
> * [X] selected hover state of the element follow design spec
> * [X] pressed state of the element follow design spec
> * [X] Responsiveness rules - the component has modifier classes for all breakpoints
> * [X] Includes Compact/Cosy/Tablet design
> * [X] RTL support
> 
> 1. The code follows fundamental-styles code standards and style
> 
> * [NA] only one top level `fd-*` class is used in the file
> * [NA] BEM naming convention is used
> * [NA] Mixins are used for repeatable code (`fd-rtl`, `fd-ellipsis`, `fd-flex`, `fd-selected`, `fd-focus`, ect.)
> * [NA] A11y support - keyboard support, screenreader support, proper ARIA attributes, etc.
> * [NA] `fd-reset()` mixin is applied to all elements
> * [NA] Variables are used, if some value is used more than twice.
> * [NA] Checked if current components can be reused, instead of having new code.
> 
> 1. Testing
> 
> * [x]  Verified all styles in IE11
> * [x]  Updated tests
> * [x] last commit message should have `[ci visual]` so it can trigger chromatic visual regression (e.g. `test: run chromatic visual regression [ci visual]`)
> 
> 1. Documentation
> 
> * [x]  Storybook documentation has been created/updated
> * [ NA] Breaking Changes [wiki](https://github.com/SAP/fundamental-styles/wiki/Breaking-Changes) has been updated in case of breaking changes.

**Comment by manu-kr** [comment]
> > Few comments. Also please check all relevant boxes in the PR description.
> > 
> > * In only the first example, `A` font appears fine in themes. The other examples don't show it correctly:
> > 
> > image
> 
> @manu-kr these comments have not been addressed.

>@kavya-b As we discussed , The color bar color shouldn't change as theme changes.

**Review by kavya-b — APPROVED** [review]
My comments have been addressed, LGTM!

**Review by InnaAtanasova — APPROVED** [review]

File: stories/rich-text-editor/rich-text-editor.stories.js [hunk 9]
```diff
@@ -0,0 +1,1158 @@
...
+				<li class="fd-list__item" role="option" tabindex="0">
+				  <span class="fd-list__title">20 pt</span>
+				</li>
+			  </ul>
+			</div>
+		  </div>
+		  <div class="fd-button-split fd-button-split--transparent fd-has-margin-right-small" role="group" aria-label="button-split">
+			<button id="cozyFontcolor" class="fd-button fd-button--transparent" aria-haspopup="true" aria-expanded="false" aria-label="button" aria-controls="t4c0o273" toggleElAttrs('t4c0o273', ['aria-hidden']);
+			toggleElAttrs('cozyFontcolor', ['aria-expanded']);>
+			  <svg class="RTE_fontColorIcon" xmlns="http://www.w3.org/2000/svg"
```
Comment (InnaAtanasova) [inline stories/rich-text-editor/rich-text-editor.stories.js:220]: please remove `RTE_fontColorIcon` class

**Comment by manu-kr** [comment]
> The only thing that bothers me is - I see i'ts svg, so if there is no way of fixing it - I'll approve this PR
> image

> Will be handled in ngx.

**Comment by InnaAtanasova** [comment]
@manu-kr Travis is failing and Chromatic is skipped

**Review by InnaAtanasova — APPROVED** [review]

**Review by mikerodonnell89 — APPROVED** [review]
Nice work!

**Pull request merged**